### DISCUSSION TITLE
How does func composition should look like

### DISCUSSION THREAD

**Discussion by Shakirov-ER** [katunilya/fundom#67] [Ideas]
Now func composition looks like:

``` py
@func
def add1(x):
    return x + 1

@func
def add2(x):
    return x+2

g = add1 << add2  # add2(add1(x))
```

I think the `<<` operator will confuse people couse `<<` use for pipelines

Mb change `<<` operator for func composition to `@` (__matmul__)

**Comment by katunilya** [comment]
Personally I decided to use byte shift operators once more because such function composition code is easily converted to direct pipeline without need in changes:

```py
@func
def add1(x: int) -> int:
   return x + 1

@func
def mul5(x: int) -> int:
   return x * 5

comp = add1 << mul5  # x -> x + 1 -> (x + 1) * 5

# as functions
@pipeline
def comp(x: int):
   return Pipe(x) << add1 << mul5
```

What I want to show is: to convert composition with `<<`/`>>` operator to conveyor one just needs to provide a correct `Pipe` in the beginning.

Also there might be performance issues if we stick to only one operator without any sync/async separation. I would do the research and give feedback on that.

**Reply by Shakirov-ER** [reply]
Ok, but put `func` back in `hof` and `carry` plz

**Reply by katunilya** [reply]
There is no `cUrry` in `pymon` (but we can try to add that, as there is not bad workaround).

Also it breaks typing when we use `func`. I want to come up to a bit different composition technic. Hope today I would bring some example.

**Comment by katunilya** [comment]
I consider the following syntax for function composition:

```py
SuggestFail = Callable[[FailSuggestion], UID | DomainError]

suggest_fail: SuggestFail = (
   compose()
   >> check_author_is_group_member  # compose_future
   >> if_ok(check_accused_is_group_member)
   >> if_ok(
      choose_ok_future()
      | compose() 
        << check_is_surrentder
        << if_ok(create_fail) 
        << if_ok(approve_fail) 
        >> if_ok(database.create)0
      | compose() << create_fail >> database.create
   )
```
